### DISCUSSION TITLE
Preferred CUE style for identifiers

### DISCUSSION THREAD

**Discussion by seh** [cuelang/cue#972] [Q&A]
I am once again interested in advice for CUE style. In particular, how do we write identifiers?

Here are some examples I’ve seen:
- Definitions are written in camel case.
- Definitions have their first letter capitalized (e.g. `#BigThing`).
- Field names are written in camel case.
- Field names have their first letter in lowercase (e.g. `thingPart`).
- **or** Field names are written in snake case (e.g. `thing_part`).

Of course, it's possible to avoid some of these decisions if all of your identifiers are made up of only one word.

JSON offers some precedent, using either snake case or lowercase-prefixed camel case for object field names. Prevailing YAML style seems to favor camel case. The prevailing Go style is camel case, forcibly using capitalization to indicate whether identifier is exported or not. In CUE, by contrast, we now have the leading underscore character rather than that capitalization toggle.

I looked at [the _unity_ project's corpus](https://github.com/cue-sh/unity/tree/main/projects), and found mixtures of these styles even within a given project. Has anyone settled on a style they can recommend?

**Comment by mpvl** [comment]
There isn't a single style yet. Part of the reason is that the style used may be forced by the underlying API that CUE intends to mode. Camel case (camelCase) seems to be most common style in use for JSON. But often this is not the case.

Originally (before `#`-style definitions), the idea was to use camelCase for fields and PascalCase for definitions. But because of the above, that didn't quite pan out all that well.

It may seem that there is more control over casing of definitions, but even that is often not the case. In some mappings, e.g. JSON Schema -> CUE, the source is mapped to definitions. This means that definitions are often camelCase as well.

Of course, we could have a recommended style and then optionally do imports converting any given style to this style. But that can be cumbersome.

There is also a matter of naming for aliases. There I think the more common style is to use PascalCase or even ALL CAPS.

> In CUE, by contrast, we now have the leading underscore character rather than that capitalization toggle.

In very early CUE we used capitalization as an export toggle. That turned out to be just too cumbersome because of the lack of standardization around JSON. Because of the clarity of this mechanism, we did adopt a leading character approach in the form of `#` and `_` as an alternative.

That said, if it were up to me, I would adopt a style a long the following lines:
1. `regularField`
2. `#Definition`
3.  `_hiddenField`
4. `ShortAlias`
5. `@attribute()`

The only ambiguity reading these is between regular fields and aliases though. So I would recommend adopting the opposed first-letter case for aliases of whatever is the style for the first letter of a regular field.

**Reply by seh** [reply]
Thank you laying this out. This matches what I've been writing recently, with one exception: I haven't been starting my definition names with a capital letter. It wouldn't be hard to switch back to that style, though.

**Reply by mpvl** [reply]
Switching the definition style may be premature, as we may settle on something different yet.

For instance, another possibility is to use `#camelCase` for struct definition, while using `#PascalCase` for constants/ fully concrete values, or vice versa.

**Reply by seh** [reply]
That's an intriguing option.

Yesterday I switched all my top-level definitions over to Pascal case. It took about ten minutes, and I found a few bugs along the way. I wasn't distinguishing then between structs and other values. Just thinking about it now, I can't decide whether Pascal case for constants and camel case for structs feels more natural to me, or the converse.

Since we don't make the distinction in Go—the language that feels most similar to writing in CUE—it doesn't supply a nudge of precedent one way or the other.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/972.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
